Add changes to expose metrics for installed knative version

savitaashture opened the pull request:
JIRA: https://jira.coreos.com/browse/SRVKS-187

**Description**:
This PR include changes to expose installed knative version information to metric and servicemonitor.

**Log**:
```
{"level":"info","ts":1566300456.1615322,"logger":"openshift","msg":"Installing Service Monitor for Operator"}
{"level":"info","ts":1566300456.631283,"logger":"openshift","msg":"Detected","gvk":"monitoring.coreos.com/v1, Kind=servicemonitor"}
{"level":"info","ts":1566300456.763179,"logger":"manifestival","msg":"Reading file","name":"deploy/resources/monitoring/operator_service_monitor.yaml"}
{"level":"info","ts":1566300456.7725942,"logger":"manifestival","msg":"Creating","name":"final/knative-serving-operator","type":"monitoring.coreos.com/v1, Kind=ServiceMonitor"}
{"level":"info","ts":1566300456.7867358,"logger":"openshift","msg":"Installing role and roleBinding for Operator"}
{"level":"info","ts":1566300456.7888649,"logger":"manifestival","msg":"Reading file","name":"deploy/resources/monitoring/role_service_monitor.yaml"}
{"level":"info","ts":1566300456.7965045,"logger":"manifestival","msg":"Creating","name":"final/knative-serving-prometheus-k8s","type":"rbac.authorization.k8s.io/v1, Kind=Role"}
{"level":"info","ts":1566300457.1032627,"logger":"manifestival","msg":"Creating","name":"final/knative-serving-prometheus-k8s","type":"rbac.authorization.k8s.io/v1, Kind=RoleBinding"}
{"level":"info","ts":1566300457.2129986,"logger":"controller_knativeserving","msg":"Install succeeded","version":"0.8.0"}
{"level":"info","ts":1566300457.2148328,"logger":"controller_knativeserving","msg":"expose metrics for installed knative serving"}
```
![Screenshot from 2019-08-20 17-00-12](https://user-images.githubusercontent.com/9441662/63343799-7fd41300-c36c-11e9-9e21-ab33d60bd8ba.png)

wtam2018 commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -556,3 +573,76 @@ func configureLogURLTemplate(u *unstructured.Unstructured) error {
+			return nil
+		}
+		log.Error(err, "Unable to create ServiceMonitor")
+		return err
```
Should we clean up the label we just created if it failed to install?

markusthoemmes replied:
I don't think we should, eventually we should succeed installing it anyway.

wtam2018 replied:
I guess the question is hpw is the label used?   Incorrect label is incorrect, right?    There is no guarantee installing is always successful eventually.

wtam2018 commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -556,3 +573,76 @@ func configureLogURLTemplate(u *unstructured.Unstructured) error {
+	if _, err := metric.CreateServiceMonitors(cfg, namespace, []*v1.Service{svcDetail}); err != nil {
+		if strings.Contains(err.Error(), "already exists") {
+			log.Info(fmt.Sprintf("ServiceMonitor already exist for %s service", svcDetail.Name))
+			return nil
```
Is it OK to use the new label we just created if monitoring has already been installed before?

markusthoemmes replied:
I think it is, adding the label should be idempotent.

markusthoemmes commented:
This is not about metrics for the instance but about metrics for the operator itself. We need to know whichever namespace the operator is installed in so we can hook that up with the ServiceMonitor etc etc.

Especially in the light of SRVKS-214 the namespace entanglement might not be clear cut.

savitaashture commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -556,3 +573,76 @@ func configureLogURLTemplate(u *unstructured.Unstructured) error {
+
+func exposeMetricToServiceMonitor(instance *servingv1alpha1.KnativeServing) error {
+	// namespace where knative-serving-operator has been installed
+	namespace, err := getNamespace()
```
instance.GetNamespace() will not give the namespace where operator installed.
And these changes are to get the metrics of operator.
So getting namespace based on NAMESPACE env var

wtam2018 replied:
Serving operator is installed in cluster scope (at least I get that when in install serving operator from OCP console/operator hub).   It looks like `instance.GetNamespace()` should give you the namespace of the target CR/Request.   https://github.com/openshift-knative/knative-serving-operator/blob/da6623f97deaa59b22878d9c1758298541dc3336/pkg/controller/knativeserving/knativeserving_controller.go#L111  Adding NAMESPACE has a lot of drawbacks.  I'd probably only use it as the last resort.

wtam2018 replied:
I don't see any easy way to get operator's namespace.    I think env var. approach is fine.

nak3 commented:
Umm... In my env I needed to creat the Role to monitor from Prometheus console, but Savita said that her test was fine. So, probably I am missing something... The code looks fine if @wtam2018's comments were addressed. (Thank you @wtam2018 !)

markusthoemmes commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -556,3 +573,76 @@ func configureLogURLTemplate(u *unstructured.Unstructured) error {
+	return ns, nil
+}
+
+func labelNamespace(namespace string) error {
```
Can we name this `addMonitoringLabelToNamespace` or something? `labelNamespace` sounds more generic than it is.

markusthoemmes commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -556,3 +573,76 @@ func configureLogURLTemplate(u *unstructured.Unstructured) error {
+	return nil
+}
+
+func getSvcDetails(ns string) *corev1.Service {
```
```suggestion
func getOperatorService(ns string) *corev1.Service {
```

? I'd also expect this to surface an error if that doesn't work. We might want to ignore a `NotFound` error only and return nil in that case.

markusthoemmes commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -556,3 +573,76 @@ func configureLogURLTemplate(u *unstructured.Unstructured) error {
+
+func getSvcDetails(ns string) *corev1.Service {
+	svc := &corev1.ServiceList{}
+	if err := api.List(context.TODO(), &client.ListOptions{Namespace: ns}, svc); err != nil {
```
Why even bother listing `api.Get(context.TODO(), ns, knative-serving-operator` should do I think?

bbrowning commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -40,25 +47,42 @@ const (
+	knativeVersion = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "knative_serving_version",
+		Help: "Installed knative serving version info",
+	}, []string{"version", "namespace", "component"})
```
This is too high of a cardinality to be useful in Prometheus. The unique combinations of these label values is potentially an unbounded set, and they should be bounded to at most 10 unique combinations as a general rule of thumb. See the CAUTION section of https://prometheus.io/docs/practices/naming/#labels

We need to rethink what labels are important here and what all the possible combinations of those label values can be.

markusthoemmes replied:
As discussed in Slack, I believe we can drop namespace and component (effectively constant anyway) and only worry about the version string. Over time, that might become an issue assuming one accumulates lots of version. However, Prometheus drops old time series after a certain amount of time (15 days by default) so I'm not sure if this is an issue really.

savitaashture replied:
So now do we need to keep only version in labels?

markusthoemmes replied:
I'd propose to go forward with keeping namespace and version, dropping component. Component is effectively already encoded in the name "knative_serving_xxx", namespace is effectively constant currently but might be more fluid in the future, so that prevents stuff from breaking and version is what we need here.

savitaashture commented:
@markusthoemmes @wtam2018 @nak3 addressed all the comments and rebased PR

nak3 commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -581,3 +599,137 @@ func installNetworkPolicies(instance *servingv1alpha1.KnativeServing) error {
+			return nil
+		}
+		log.Error(err, "failed to create ServiceMonitor so removing added label from namespace")
+		if err = removeMonitoringLabelFromNamespace(namespace); err != nil {
```
If other ServiceMonitor is deployed and the label had already been existed, this func removes the label? I wonder that we should remove the label, only when `addMonitoringLabelToNamespace` newly created it.

savitaashture replied:
`addMonitoringLabelToNamespace` will add label to the namespace where operator is installed so in that case `removeMonitoringLabelFromNamespace` will remove if there is failure in servicemonitor creation 

In this scenario no other service monitor will be created.

nak3 replied:
Ah, no I mean, for example, if you deployed serving-operator on `openshift-monitoring` ns and it failed, the label will be removed. But `openshift-monitoring` ns still needs the label because other ServiceMonitors use it. (please refer to `$ oc get servicemonitor -n openshift-monitoring`)
This scenario does not happen?

markusthoemmes replied:
I'd ditch the cleanup logic. The label doesn't hurt, does it?

nak3 replied:
Correct. The label does not hurt. So I agree that we ditch the clean up logic.

markusthoemmes commented on deploy/operator.yaml:
```diff
@@ -15,13 +15,18 @@ spec:
         - name: knative-serving-operator
-          image: quay.io/openshift-knative/knative-serving-operator:v0.8.0
+          #image: quay.io/openshift-knative/knative-serving-operator:v0.8.0
+          image: savita3020/knative-serving-operator:0.8.1
```
This shouldn't be checked in.

markusthoemmes commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -581,3 +599,137 @@ func installNetworkPolicies(instance *servingv1alpha1.KnativeServing) error {
+}
+
+func createRole(ns string) error {
+	role := &rbac.Role{}
```
Why not adding all of the fields in here directly instead of mutating it in the following lines?

markusthoemmes commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -581,3 +599,137 @@ func installNetworkPolicies(instance *servingv1alpha1.KnativeServing) error {
+}
+
+func createRoleBinding(ns string) error {
+	roleBinding := &rbac.RoleBinding{}
```
Same here regarding building the object in place.

markusthoemmes commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -581,3 +599,137 @@ func installNetworkPolicies(instance *servingv1alpha1.KnativeServing) error {
+		Verbs:     []string{"get", "list", "watch"},
+	}
+	role.Rules = append(role.Rules, rules)
+	err := api.Create(context.TODO(), role)
```
```suggestion
	return api.Create(context.TODO(), role)
```

markusthoemmes commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -581,3 +599,137 @@ func installNetworkPolicies(instance *servingv1alpha1.KnativeServing) error {
+	roleBinding.Namespace = ns
+	roleBinding.RoleRef = roleRef
+	roleBinding.Subjects = append(roleBinding.Subjects, subjects)
+	err := api.Create(context.TODO(), roleBinding)
```
```suggestion
	return api.Create(context.TODO(), roleBinding)
```

markusthoemmes commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -581,3 +599,137 @@ func installNetworkPolicies(instance *servingv1alpha1.KnativeServing) error {
+	if err := api.Get(context.TODO(), client.ObjectKey{Name: namespace}, ns); err != nil {
+		return err
+	}
+	if len(ns.Labels) == 0 {
```
```suggestion
	if ns.Labels == nil {
```

markusthoemmes commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -581,3 +599,137 @@ func installNetworkPolicies(instance *servingv1alpha1.KnativeServing) error {
+	return nil
+}
+
+func removeMonitoringLabelFromNamespace(namespace string) error {
```
As mentioned above, I don't think we need this.

markusthoemmes commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -581,3 +599,137 @@ func installNetworkPolicies(instance *servingv1alpha1.KnativeServing) error {
+	// Get service knative-serving-operator in order to create service monitor
+	svcDetail, err := getOperatorService(namespace)
+	if err != nil {
+		if strings.Contains(err.Error(), "not found") {
```
Is this an actual apimachinery error? If so, please use https://godoc.org/k8s.io/apimachinery/pkg/api/errors#IsNotFound

markusthoemmes commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -581,3 +599,137 @@ func installNetworkPolicies(instance *servingv1alpha1.KnativeServing) error {
+
+	// CreateServiceMonitors creates ServiceMonitors objects based on Service objects.
+	if _, err := metric.CreateServiceMonitors(cfg, namespace, []*v1.Service{svcDetail}); err != nil {
+		if strings.Contains(err.Error(), "already exists") {
```
This should also be an apimachinery error, please use https://godoc.org/k8s.io/apimachinery/pkg/api/errors#IsAlreadyExists

markusthoemmes commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -581,3 +599,137 @@ func installNetworkPolicies(instance *servingv1alpha1.KnativeServing) error {
+	// getOperatorNamespace return namespace where knative-serving-operator has been installed
+	namespace, err := getOperatorNamespace()
+	if err != nil {
+		log.Info(fmt.Sprintf("empty %s because operator dont want to expose metric or running locally", knativeServingInstalledNamespace))
```
```suggestion
		log.Info("no namespace defined, skipping ServiceMonitor installation for the operator")
```

markusthoemmes commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -581,3 +599,137 @@ func installNetworkPolicies(instance *servingv1alpha1.KnativeServing) error {
+	return nil
+}
+
+func createRole(ns string) error {
```
Please be a bit more explicit wrt. the name of these two functions.

markusthoemmes commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -581,3 +598,121 @@ func installNetworkPolicies(instance *servingv1alpha1.KnativeServing) error {
+	if _, err := metric.CreateServiceMonitors(cfg, namespace, []*v1.Service{svcDetail}); err != nil {
+		if errors.IsAlreadyExists(err) {
+			log.Info(fmt.Sprintf("ServiceMonitor already exist for %s service", svcDetail.Name))
+			return nil
```
This `return nil` is actually wrong I think? We want to try and apply the role and rolebinding below as well. They'll need the same treatment wrt. ignoring `IsAlreadyExists` errors I think?

markusthoemmes commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -581,3 +598,121 @@ func installNetworkPolicies(instance *servingv1alpha1.KnativeServing) error {
+	return nil
+}
+
+func createRoleForServiceMonitor(ns string) error {
```
These two functions need to be idempotent to correctly pass the reconcile flow.

That usually goes like this:

```golang
obj, err := api.Get(...)
if errors.IsNotFound(err) {
  obj := constructObj()
  return api.Create(obj)
} else if err != nil {
  return err
} else {
  obj := constructObj()
  return api.Update(obj)
}
```

Have you thought about creating a similar manifest to that we already have in `/deploy/resources/monitoring` and run a transformer over it that adds the namespace at runtime? That'd make this a tad bit easier as that handles all the edge-cases for you I believe.

savitaashture replied:
@markusthoemmes Thank you 
I have modified according to the comment.

nak3 commented on deploy/resources/monitoring/operator_service_monitor.yaml:
```diff
@@ -0,0 +1,45 @@
+    matchLabels:
+      name: knative-serving-operator
+---
+apiVersion: rbac.authorization.k8s.io/v1
```
This Role & RoleBinding are duplicated with `service_monitor.yaml`'s. Should we create the separate file and share it between `installServiceMonitor` and `exposeMetricToServiceMonitor`?

nak3 commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -581,3 +565,77 @@ func installNetworkPolicies(instance *servingv1alpha1.KnativeServing) error {
+		log.Error(err, "Unable to create ServiceMonitor install manifest")
+		return err
+	}
+	transforms := []mf.Transformer{mf.InjectOwner(instance)}
```
I am wondering that we would clean up the existing Role&RoleBinding whenever we uninstall the serving-operator. I mean that the same problem with [this](https://github.com/openshift-knative/knative-serving-operator/pull/49) happened on Role&Rolebiding. If so, I think we should change the Role&Role name (=prometheus-k8s).

markusthoemmes commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -581,3 +565,77 @@ func installNetworkPolicies(instance *servingv1alpha1.KnativeServing) error {
+		return nil
+	}
+	log.Info("exposing metric for installed knative version")
+	knativeVersion.WithLabelValues(version.Version, namespace).Set(float64(time.Now().Unix()))
```
This namespace actually shouldn't be the operator's namespace, should it? I'd move this line to the beginning of the reconcile loop actually. It's also not openshift specific so let's move it into the general path?

savitaashture replied:
Then which namespace?
I feel we can remove `namespace` also because by default metric add `exported_namespace` as show in the screenshot.

markusthoemmes replied:
yeah I think it's fine either way. It'd be premature optimization for a maybe future world where you can have more than one knativeServing per cluster and have different versions for them... but that's faaaar out and we can always change later.

savitaashture replied:
To make it general planning to add changes to expose metric once after 
`log.Info("All deployments are available")`

markusthoemmes replied:
I'd actually expose it in any case. Healthyness stats should be written separately.

nak3 commented on pkg/controller/knativeserving/openshift/openshift.go:
```diff
@@ -50,7 +55,7 @@ var (
 		Transformers: []mf.Transformer{ingress, egress, deploymentController, annotateAutoscalerService, augmentAutoscalerDeployment, addCaBundleToApiservice, configureLogURLTemplate},
 		PreInstalls:  []common.Extender{addUsersToSCCs, installNetworkPolicies, ensureMaistra, caBundleConfigMap},
-		PostInstalls: []common.Extender{ensureOpenshiftIngress, installServiceMonitor},
+		PostInstalls: []common.Extender{ensureOpenshiftIngress, installServiceMonitor, exposeMetricToServiceMonitor},
```
We should change the function names `installServiceMonitor` and/or `exposeMetricToServiceMonitor`? These two functions do similar things, but the names look inconsistent.

savitaashture replied:
two function do the same thing but for different namespace.
I think its good to change the function names and refactor code according to that WDYT?

nak3 replied:
Yes, I agree :+1:

savitaashture commented:
Closing
Addressed all the comments and raised same PR against TP1 branch
https://github.com/openshift-knative/knative-serving-operator/pull/54

Thank you
